# Data Cleaning Project

#### Prepared by Faiz Yah, 19th February 2024, KL Malaysia

When we first receive a new fresh set of data, sometimes it arrives in a form that is uncleaned and dirty, hence we need to clean the data first. Uncleaned data could potentially lead to incorrect analysis and outcome, which will negatively impact the entire business' productivity.

Data cleaning, also known as data cleansing, generally includes removing unnecessary columns/rows, removing duplicated values/null values/outliers, replacing values, fixing formats, to name a few.

This process is one of the crucial steps taken before analysing the data. A well-cleaned data increases the quality and efficiency of the data, ensuring better decision-making while simultaneously saves resources.



## 1. Introduction of Project

The main focus of the project is to **clean raw data.**

The data set records the **numbers and stats** of the football players in the **game FIFA 21**, taken from [Kaggle](https://www.kaggle.com/datasets/yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring).

The data set is **rather messy** (or Messi) in its raw state and requires thorough cleaning.



## 2. Execution of Project

The data cleaning procedures carried out for this project are as follows:

1. Download and import the _csv._ file
2. Check information about the dataset (Amount of columns and their respective data type)
3. Browse through the dataset (To gain first impression and better understanding)
4. Remove unnecessary columns
5. Modify and extract data of column 
	5.1 Convert height from inches to cm
	5.2 Convert weight from pounds to kilograms 
6. Split column into 2 separate columns
7. Convert data type of column
8. Remove unwanted rows
9. Extract and convert 'M - million' and 'K - thousand' strings into int data type
10. Remove unnecessary columns 
11. Export the cleaned data into _.csv_ or _.json_ file

### 1. Import the _csv._ file

In [1]:
import pandas as pd

df = pd.read_csv('fifa21_raw_data.csv')

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (76) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


### 2. Check the information of the data set

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   photoUrl          18979 non-null  object
 1   LongName          18979 non-null  object
 2   playerUrl         18979 non-null  object
 3   Nationality       18979 non-null  object
 4   Positions         18979 non-null  object
 5   Name              18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Team & Contract   18979 non-null  object
 10  ID                18979 non-null  int64 
 11  Height            18979 non-null  object
 12  Weight            18979 non-null  object
 13  foot              18979 non-null  object
 14  BOV               18979 non-null  int64 
 15  BP                18979 non-null  object
 16  Growth            18979 non-null  int64 
 17  Joined      

I noticed that there are a total of **77 columns**, majority of them are **the stats of the players**.

In [3]:
df.shape

(18979, 77)

We can also check using the _.shape_ function, and the result shows that there are **18979 rows and 77 columns**.

### 3. Browse through the 77 columns

In [4]:
pd.set_option('display.max_columns', None) # Apply this code to be able to browse through all 77 columns.

df.head()

photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                       LongName  \
0                  Lionel Messi   
1  C. Ronaldo dos Santos Aveiro   
2                     Jan Oblak   
3               Kevin De Bruyne   
4    Neymar da Silva Santos Jr.   

                                           playerUrl Nationality Positions  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina  RW ST CF   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal     ST LW   
2  http://sofifa.com/player/200389/jan-oblak/210005/    Slovenia        GK   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium    CAM CM   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil    LW CAM   

                Name  Age  ↓OVA  POT  \
0           L. Messi   33    93   93   
1  Cristiano Ronaldo   35    92   92   
2           J. Oblak   27    91   93   
3       K. De Bruyne   29    91   91   
4          Neymar Jr   28    91   91   

                                Team & Contract      ID Height  Weight   foot  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n  158023   5'7"  159lbs   Left   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   20801   6'2"  183lbs  Right   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n  200389   6'2"  192lbs  Right   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  192985  5'11"  154lbs  Right   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n  190871   5'9"  150lbs  Right   

   BOV   BP  Growth        Joined Loan Date End   Value   Wage Release Clause  \
0   93   RW       0   Jul 1, 2004           NaN  €67.5M  €560K        €138.4M   
1   92   ST       0  Jul 10, 2018           NaN    €46M  €220K         €75.9M   
2   91   GK       2  Jul 16, 2014           NaN    €75M  €125K        €159.4M   
3   91  CAM       0  Aug 30, 2015           NaN    €87M  €370K          €161M   
4   91   LW       0   Aug 3, 2017           NaN    €90M  €270K        €166.5M   

   Attacking  Crossing  Finishing  Heading Accuracy  Short Passing  Volleys  \
0        429        85         95                70             91       88   
1        437        84         95                90             82       86   
2         95        13         11                15             43       13   
3        407        94         82                55             94       82   
4        408        85         87                62             87       87   

   Skill  Dribbling  Curve  FK Accuracy  Long Passing  Ball Control  Movement  \
0    470         96     93           94            91            96       451   
1    414         88     81           76            77            92       431   
2    109         12     13           14            40            30       307   
3    441         88     85           83            93            92       398   
4    448         95     88           89            81            95       453   

   Acceleration  Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  \
0            91            80       91         94       95    389          86   
1            87            91       87         95       71    444          94   
2            43            60       67         88       49    268          59   
3            77            76       78         91       76    408          91   
4            94            89       96         91       83    357          80   

   Jumping  Stamina  Strength  Long Shots  Mentality  Aggression  \
0       68       72        69          94        347          44   
1       95       84        78          93        353          63   
2       78       41        78          12        140          34   
3       63       89        74          91        408          76   
4       62       81        5

Browsing through the 77 columns is essential as it provides an initial impression and better understanding of the data set. It allows us to identify what consist in the data sets, which data **should be kept, while which should be removed**.



### 4. Remove unnecessary columns

In [5]:
# Remove all columns between column index 22 to 76

df.drop(df.iloc[:, 22:77], inplace=True, axis=1) 

After the browse through, my verdicts are that I am not very interested with the stats of each players, so I will drop these columns with the player stats.

For the column removal, instead of inputing the name of each column individually, I can use the **column index**. The column indexes can be identified from the df.info() above.

For this case, the column index I would like to remove are from column index 22 to 76.

In [6]:
df.head()

photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                       LongName  \
0                  Lionel Messi   
1  C. Ronaldo dos Santos Aveiro   
2                     Jan Oblak   
3               Kevin De Bruyne   
4    Neymar da Silva Santos Jr.   

                                           playerUrl Nationality Positions  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina  RW ST CF   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal     ST LW   
2  http://sofifa.com/player/200389/jan-oblak/210005/    Slovenia        GK   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium    CAM CM   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil    LW CAM   

                Name  Age  ↓OVA  POT  \
0           L. Messi   33    93   93   
1  Cristiano Ronaldo   35    92   92   
2           J. Oblak   27    91   93   
3       K. De Bruyne   29    91   91   
4          Neymar Jr   28    91   91   

                                Team & Contract      ID Height  Weight   foot  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n  158023   5'7"  159lbs   Left   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   20801   6'2"  183lbs  Right   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n  200389   6'2"  192lbs  Right   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  192985  5'11"  154lbs  Right   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n  190871   5'9"  150lbs  Right   

   BOV   BP  Growth        Joined Loan Date End   Value   Wage Release Clause  
0   93   RW       0   Jul 1, 2004           NaN  €67.5M  €560K        €138.4M  
1   92   ST       0  Jul 10, 2018           NaN    €46M  €220K         €75.9M  
2   91   GK       2  Jul 16, 2014           NaN    €75M  €125K        €159.4M  
3   91  CAM       0  Aug 30, 2015           NaN    €87M  €370K          €161M  
4   91   LW       0   Aug 3, 2017           NaN    €90M  €270K        €166.5M

It seems that I have succesfully dropped the columns of the stats of the player.

Next, I would like to drop the "Url" columns, I can use the similar coding as above.

In [7]:
df.drop(df.columns[[0,2]], inplace=True, axis=1) 

df.head()

LongName Nationality Positions               Name  Age  \
0                  Lionel Messi   Argentina  RW ST CF           L. Messi   33   
1  C. Ronaldo dos Santos Aveiro    Portugal     ST LW  Cristiano Ronaldo   35   
2                     Jan Oblak    Slovenia        GK           J. Oblak   27   
3               Kevin De Bruyne     Belgium    CAM CM       K. De Bruyne   29   
4    Neymar da Silva Santos Jr.      Brazil    LW CAM          Neymar Jr   28   

   ↓OVA  POT                               Team & Contract      ID Height  \
0    93   93         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n  158023   5'7"   
1    92   92             \n\n\n\nJuventus\n2018 ~ 2022\n\n   20801   6'2"   
2    91   93      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n  200389   6'2"   
3    91   91      \n\n\n\nManchester City\n2015 ~ 2023\n\n  192985  5'11"   
4    91   91  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n  190871   5'9"   

   Weight   foot  BOV   BP  Growth        Joined Loan Date End   Value   Wage  \
0  159lbs   Left   93   RW       0   Jul 1, 2004           NaN  €67.5M  €560K   
1  183lbs  Right   92   ST       0  Jul 10, 2018           NaN    €46M  €220K   
2  192lbs  Right   91   GK       2  Jul 16, 2014           NaN    €75M  €125K   
3  154lbs  Right   91  CAM       0  Aug 30, 2015           NaN    €87M  €370K   
4  150lbs  Right   91   LW       0   Aug 3, 2017           NaN    €90M  €270K   

  Release Clause  
0        €138.4M  
1         €75.9M  
2        €159.4M  
3          €161M  
4        €166.5M

Next, I would like to drop the LongName, ID, Growth, Load Date End, Release Clause columns. I can use the similar code as above.

In [8]:
df.info()  # To identify the index of the columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   LongName         18979 non-null  object
 1   Nationality      18979 non-null  object
 2   Positions        18979 non-null  object
 3   Name             18979 non-null  object
 4   Age              18979 non-null  int64 
 5   ↓OVA             18979 non-null  int64 
 6   POT              18979 non-null  int64 
 7   Team & Contract  18979 non-null  object
 8   ID               18979 non-null  int64 
 9   Height           18979 non-null  object
 10  Weight           18979 non-null  object
 11  foot             18979 non-null  object
 12  BOV              18979 non-null  int64 
 13  BP               18979 non-null  object
 14  Growth           18979 non-null  int64 
 15  Joined           18979 non-null  object
 16  Loan Date End    1013 non-null   object
 17  Value            18979 non-null

In [9]:
df.drop(df.columns[[0,8,14,16,19]], inplace=True, axis=1) 

df.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                                Team & Contract Height  Weight   foot  BOV  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n   5'7"  159lbs   Left   93   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   6'2"  183lbs  Right   92   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n   6'2"  192lbs  Right   91   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  5'11"  154lbs  Right   91   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n   5'9"  150lbs  Right   91   

    BP        Joined   Value   Wage  
0   RW   Jul 1, 2004  €67.5M  €560K  
1   ST  Jul 10, 2018    €46M  €220K  
2   GK  Jul 16, 2014    €75M  €125K  
3  CAM  Aug 30, 2015    €87M  €370K  
4   LW   Aug 3, 2017    €90M  €270K

Great, I have removed all the columns that are not required. We went from having **77 columns, to only 16 columns.**

### 5.1 Modify the height column

Next, I will be converting the **height and weight** into **float type** which will allow us to apply them to **aggregrate functions** later.

In [10]:
# To convert the Height in inches into Height_cm which allows easier calculation for aggregration functions

df['Height_cm'] = (df['Height']
 .str.extract(r'(\d+)\'\s*(\d+)"')
 .astype(float)
 .mul([12*2.54, 2.54])
 .sum(axis=1)
 )

# To rearrange the columns, so that the column "Height_cm" sits beside "Height"
df = df[['Nationality', 'Positions', 'Name', 'Age', '↓OVA', 'POT',
       'Team & Contract', 'Height','Height_cm', 'Weight', 'foot', 'BOV', 'BP',
       'Joined', 'Value', 'Wage']]   

df.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                                Team & Contract Height  Height_cm  Weight  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n   5'7"     170.18  159lbs   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   6'2"     187.96  183lbs   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n   6'2"     187.96  192lbs   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  5'11"     180.34  154lbs   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n   5'9"     175.26  150lbs   

    foot  BOV   BP        Joined   Value   Wage  
0   Left   93   RW   Jul 1, 2004  €67.5M  €560K  
1  Right   92   ST  Jul 10, 2018    €46M  €220K  
2  Right   91   GK  Jul 16, 2014    €75M  €125K  
3  Right   91  CAM  Aug 30, 2015    €87M  €370K  
4  Right   91   LW   Aug 3, 2017    €90M  €270K

A success! We have now added in the **_Height_cm_** column, which converts the height from inches to cm, and this column is also in the **float** type.

### 5.2 Modify the weight column

In [11]:
# To strip the numerical value from the lbs in the weight columns, then convert it to kilogram
df['Weight_kg'] = df['Weight'].str.extract(r'(\d+)').astype(float).mul(0.453592).round(2)

# To rearrange the columns, so that the column "Height_cm" sits beside "Height"
df = df[['Nationality', 'Positions', 'Name', 'Age', '↓OVA', 'POT',
       'Team & Contract', 'Height','Height_cm', 'Weight', 'Weight_kg', 'foot', 'BOV', 'BP',
       'Joined', 'Value', 'Wage']]   

df.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                                Team & Contract Height  Height_cm  Weight  \
0         \n\n\n\nFC Barcelona\n2004 ~ 2021\n\n   5'7"     170.18  159lbs   
1             \n\n\n\nJuventus\n2018 ~ 2022\n\n   6'2"     187.96  183lbs   
2      \n\n\n\nAtlético Madrid\n2014 ~ 2023\n\n   6'2"     187.96  192lbs   
3      \n\n\n\nManchester City\n2015 ~ 2023\n\n  5'11"     180.34  154lbs   
4  \n\n\n\nParis Saint-Germain\n2017 ~ 2022\n\n   5'9"     175.26  150lbs   

   Weight_kg   foot  BOV   BP        Joined   Value   Wage  
0      72.12   Left   93   RW   Jul 1, 2004  €67.5M  €560K  
1      83.01  Right   92   ST  Jul 10, 2018    €46M  €220K  
2      87.09  Right   91   GK  Jul 16, 2014    €75M  €125K  
3      69.85  Right   91  CAM  Aug 30, 2015    €87M  €370K  
4      68.04  Right   91   LW   Aug 3, 2017    €90M  €270K

Similar to the _height_ column, we now created a new **_Weight_kg_** column that converts the players weight from pounds to kg, and the column is in the **float** type.

### 6.1 Split the 'Team & Contract' column 

The team and contract year are merged into one column, is is required to split them into 2 seperate columns.

But first, we will have to clean and remove the repeated "\n" strings.

In [12]:
# To convert the column into "string" type
df['Team & Contract'] = df['Team & Contract'].astype('str') 

# To remove the repeated "/n" string
df['Team & Contract'] = df['Team & Contract'].str.replace('\n', '')

df.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                  Team & Contract Height  Height_cm  Weight  Weight_kg   foot  \
0         FC Barcelona2004 ~ 2021   5'7"     170.18  159lbs      72.12   Left   
1             Juventus2018 ~ 2022   6'2"     187.96  183lbs      83.01  Right   
2      Atlético Madrid2014 ~ 2023   6'2"     187.96  192lbs      87.09  Right   
3      Manchester City2015 ~ 2023  5'11"     180.34  154lbs      69.85  Right   
4  Paris Saint-Germain2017 ~ 2022   5'9"     175.26  150lbs      68.04  Right   

   BOV   BP        Joined   Value   Wage  
0   93   RW   Jul 1, 2004  €67.5M  €560K  
1   92   ST  Jul 10, 2018    €46M  €220K  
2   91   GK  Jul 16, 2014    €75M  €125K  
3   91  CAM  Aug 30, 2015    €87M  €370K  
4   91   LW   Aug 3, 2017    €90M  €270K

We have succesfully **remove the repeated '\n'** strings by replacing it with '' or in simple words: nothing.

There are 2 methods to split the columns, both methods use string slicing. The 2nd method uses a for loop. 

In [13]:
# METHOD 1

df['Team_Name'] = df['Team & Contract'].str[:-11] # To obtain only the first 11 string, starting from the end
df['Contract_Year'] = df['Team & Contract'].str[-11:] # To obtain the remaining strings after the 11th string, starting from the end

# To rearrange the columns
df = df[['Nationality', 'Positions', 'Name', 'Age', '↓OVA', 'POT',
       'Team & Contract', 'Team_Name', 'Contract_Year', 'Height','Height_cm', 'Weight', 'Weight_kg', 'foot', 'BOV', 'BP',
       'Joined', 'Value', 'Wage']]  

In [ ]:
# METHOD 2 
Team = []
Year=[]

for i in range(len(df['Team & Contract'])):
    Team.append(str(df['Team & Contract'][i][:-11])) # To obtain the team name
    Year.append(str(df['Team & Contract'][i][-11:])) # To obtain the contract year
    
df['Team_Name'] = Team
df['Contract_Year'] = Year

In [15]:
df.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                  Team & Contract            Team_Name Contract_Year Height  \
0         FC Barcelona2004 ~ 2021         FC Barcelona   2004 ~ 2021   5'7"   
1             Juventus2018 ~ 2022             Juventus   2018 ~ 2022   6'2"   
2      Atlético Madrid2014 ~ 2023      Atlético Madrid   2014 ~ 2023   6'2"   
3      Manchester City2015 ~ 2023      Manchester City   2015 ~ 2023  5'11"   
4  Paris Saint-Germain2017 ~ 2022  Paris Saint-Germain   2017 ~ 2022   5'9"   

   Height_cm  Weight  Weight_kg   foot  BOV   BP        Joined   Value   Wage  
0     170.18  159lbs      72.12   Left   93   RW   Jul 1, 2004  €67.5M  €560K  
1     187.96  183lbs      83.01  Right   92   ST  Jul 10, 2018    €46M  €220K  
2     187.96  192lbs      87.09  Right   91   GK  Jul 16, 2014    €75M  €125K  
3     180.34  154lbs      69.85  Right   91  CAM  Aug 30, 2015    €87M  €370K  
4     175.26  150lbs      68.04  Right   91   LW   Aug 3, 2017    €90M  €270K

We have **succesfully split** the columns as wanted.

For better clarity, we also arranged the columns, allowing related columns to be placed beside each other.

### 6.2 Split the Contract_Year column

In [16]:
# To create 2 new columns that split the year into "start year" and "end year"
df['Start_Year'] = df['Contract_Year'].str[:4] 
df['End_Year'] = df['Contract_Year'].str[-4:] 

# To rearrange the columns
df = df[['Nationality', 'Positions', 'Name', 'Age', '↓OVA', 'POT',
       'Team & Contract', 'Team_Name', 'Contract_Year', 'Start_Year', 'End_Year', 
         'Height','Height_cm', 'Weight', 'Weight_kg', 'foot', 'BOV', 'BP',
       'Joined', 'Value', 'Wage']]   


df.head()

Nationality Positions               Name  Age  ↓OVA  POT  \
0   Argentina  RW ST CF           L. Messi   33    93   93   
1    Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2    Slovenia        GK           J. Oblak   27    91   93   
3     Belgium    CAM CM       K. De Bruyne   29    91   91   
4      Brazil    LW CAM          Neymar Jr   28    91   91   

                  Team & Contract            Team_Name Contract_Year  \
0         FC Barcelona2004 ~ 2021         FC Barcelona   2004 ~ 2021   
1             Juventus2018 ~ 2022             Juventus   2018 ~ 2022   
2      Atlético Madrid2014 ~ 2023      Atlético Madrid   2014 ~ 2023   
3      Manchester City2015 ~ 2023      Manchester City   2015 ~ 2023   
4  Paris Saint-Germain2017 ~ 2022  Paris Saint-Germain   2017 ~ 2022   

  Start_Year End_Year Height  Height_cm  Weight  Weight_kg   foot  BOV   BP  \
0       2004     2021   5'7"     170.18  159lbs      72.12   Left   93   RW   
1       2018     2022   6'2"     187.96  183lbs      83.01  Right   92   ST   
2       2014     2023   6'2"     187.96  192lbs      87.09  Right   91   GK   
3       2015     2023  5'11"     180.34  154lbs      69.85  Right   91  CAM   
4       2017     2022   5'9"     175.26  150lbs      68.04  Right   91   LW   

         Joined   Value   Wage  
0   Jul 1, 2004  €67.5M  €560K  
1  Jul 10, 2018    €46M  €220K  
2  Jul 16, 2014    €75M  €125K  
3  Aug 30, 2015    €87M  €370K  
4   Aug 3, 2017    €90M  €270K

Great, we now have the beginning year and the end year of the contract in 2 seperate columns.

### 7. Convert the Year columns into int
### and
### 8. Remove unwanted rows

Next, I want to convert the **_Start_Year_** and **_End_Year_** columns into **int** data type.

In [17]:
# To remove empty spaces
df['Start_Year'] = df['Start_Year'].replace(" ", "")

# To convert data type of column to 'int32'
df['Start_Year'] = df['Start_Year'].astype('int32') 

ValueError: invalid literal for int() with base 10: ' Bra'

However there is an error, it seems like there is a **string ' Bra'** in some of the rows. 

Let's check which rows contains this string ' Bra'.

In [18]:
# To check rows that contain ' Bra'

df[df['Start_Year'] == ' Bra']

Nationality Positions            Name  Age  ↓OVA  POT Team & Contract  \
288      Brazil     LB LM  Welington Dano   20    81   81      BrazilFree   
292      Brazil    CB CDM  Juiano Mestres   24    81   81      BrazilFree   

    Team_Name Contract_Year Start_Year End_Year Height  Height_cm  Weight  \
288              BrazilFree        Bra     Free  5'10"     177.80  152lbs   
292              BrazilFree        Bra     Free  5'11"     180.34  181lbs   

     Weight_kg   foot  BOV  BP       Joined Value Wage  
288      68.95   Left   81  LB  Jan 1, 2019    €0   €0  
292      82.10  Right   81  CB  Jan 1, 2019    €0   €0

It seems like there are **2 rows with the string ' Bra'** instead of an actual year, we need to remove these rows.

In [19]:
# To remove the rows containing the string ' Bra'

for x in df.index:
    if df.loc[x, "Start_Year"]== ' Bra':
        df.drop(x, inplace=True)

df[df['Start_Year'] == ' Bra'].count()

Nationality        0
Positions          0
Name               0
Age                0
↓OVA               0
POT                0
Team & Contract    0
Team_Name          0
Contract_Year      0
Start_Year         0
End_Year           0
Height             0
Height_cm          0
Weight             0
Weight_kg          0
foot               0
BOV                0
BP                 0
Joined             0
Value              0
Wage               0
dtype: int64

Alright the invalid rows have been removed, and we can reconfirm by counting and making sure the sum is 0.

In [20]:
df['Start_Year'] = df['Start_Year'].astype('int') 

ValueError: invalid literal for int() with base 10: 'Ecua'

Ah, the same **string invalidity** occurs again. This means that there are **more rows** in the _Year_ columns that **doesn't  have an actual year**, but instead has **some kind of string.**

Now let's see how many rows contain these invalid strings.

In [21]:
# To check rows that does not start with '2'

df[~df['Start_Year'].str.startswith("2")]  

Nationality Positions            Name  Age  ↓OVA  POT  \
205         Wales        RW         G. Bale   30    83   83   
250      Portugal    CDM CM  Danilo Pereira   28    82   82   
257        France        GK       A. Areola   27    82   86   
299       Uruguay    CDM CM     L. Torreira   24    81   85   
305         Italy  RM RW ST     M. Politano   26    81   81   
...           ...       ...             ...  ...   ...  ...   
18497   Australia        ST       M. Najjar   20    52   65   
18569    China PR        CM     Deng Yubiao   23    51   60   
18580     England  CM RW LW         A. Read   20    51   63   
18638    China PR    CAM LM         Chen Pu   23    51   60   
18696    China PR  CM LM LB    Chen Zhechao   25    51   56   

                                         Team & Contract  \
205                Tottenham HotspurJun 30, 2021 On Loan   
250              Paris Saint-GermainJun 30, 2021 On Loan   
257                           FulhamJun 30, 2021 On Loan   
299                  Atlético MadridJun 30, 2021 On Loan   
305                           NapoliJun 30, 2021 On Loan   
...                                                  ...   
18497                   Macarthur FCAug 31, 2021 On Loan   
18569  Shijiazhuang Ever Bright F.C.Dec 31, 2020 On Loan   
18580                      StevenageJun 30, 2021 On Loan   
18638  Shijiazhuang Ever Bright F.C.Dec 31, 2020 On Loan   
18696               Guangzhou R&F FCDec 31, 2020 On Loan   

                                    Team_Name Contract_Year Start_Year  \
205                Tottenham HotspurJun 30, 2   021 On Loan       021    
250              Paris Saint-GermainJun 30, 2   021 On Loan       021    
257                           FulhamJun 30, 2   021 On Loan       021    
299                  Atlético MadridJun 30, 2   021 On Loan       021    
305                           NapoliJun 30, 2   021 On Loan       021    
...                                       ...           ...        ...   
18497                   Macarthur FCAug 31, 2   021 On Loan       021    
18569  Shijiazhuang Ever Bright F.C.Dec 31, 2   020 On Loan       020    
18580                      StevenageJun 30, 2   021 On Loan       021    
18638  Shijiazhuang Ever Bright F.C.Dec 31, 2   020 On Loan       020    
18696               Guangzhou R&F FCDec 31, 2   020 On Loan       020    

      End_Year Height  Height_cm  Weight  Weight_kg   foot  BOV   BP  \
205       Loan   6'1"     185.42  181lbs      82.10   Left   83   RW   
250       Loan   6'2"     187.96  183lbs      83.01  Right   82  CDM   
257       Loan   6'5"     195.58  207lbs      93.89  Right   82   GK   
299       Loan   5'5"     165.10  132lbs      59.87  Right   82  CDM   
305       Loan   5'7"     170.18  148lbs      67.13   Left   81   RM   
...        ...    ...        ...     ...        ...    ...  ...  ...   
18497     Loan   6'0"     182.88  159lbs      72.12   Left   54   ST   
18569     Loan  5'11"     180.34  143lbs      64.86  Right   55  CAM   
18580     Loan  5'10"     177.80  141lbs      63.96   Left   55  CAM   
18638     Loan  5'10"     177.80  150lbs      68.04  Right   53  CAM   
18696     Loan  5'10"     177.80  154lbs      69.85   Left   59   CB   

             Joined   Value   Wage  
205     Sep 1, 2013  €24.5M  €130K  
250     Jul 2, 2015    €20M   €92K  
257     Jul 1, 2010    €20M   €93K  
299    Jul 12, 2018    €21M   €49K  
305     Jul 1, 2019    €19M   €74K  
...             ...     ...    ...  
18497  Feb 13, 2018   €100K   €500  
18569  Jul 14, 2017    €60K    €1K  
18580  Jul 12, 2019    €70K    €1K  
18638  Jul 25, 2019    €60K    €1K  
18696   Jan 1, 2016    €50K    €2K  

[1250 rows x 21 columns]

From the output, we can identify that there are actually **1250 rows** that **doesn't have an actual year** written in the "Year" columns.

So we have to remove these invalid rows.

In [22]:
# We remove the invalid rows.

# We achieve it by saving only the valid rows.
df = df[df['Start_Year'].str.startswith("2")]

df

Nationality Positions               Name  Age  ↓OVA  POT  \
0       Argentina  RW ST CF           L. Messi   33    93   93   
1        Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2        Slovenia        GK           J. Oblak   27    91   93   
3         Belgium    CAM CM       K. De Bruyne   29    91   91   
4          Brazil    LW CAM          Neymar Jr   28    91   91   
...           ...       ...                ...  ...   ...  ...   
18974    China PR        CB     Zhang Mengxuan   21    47   52   
18975     England        ST        V. Da Silva   17    47   67   
18976    China PR        CB             Xia Ao   21    47   55   
18977     England        CM           B. Hough   17    47   67   
18978     Bolivia       CDM          M. Flores   19    47   63   

                                      Team & Contract  \
0                             FC Barcelona2004 ~ 2021   
1                                 Juventus2018 ~ 2022   
2                          Atlético Madrid2014 ~ 2023   
3                          Manchester City2015 ~ 2023   
4                      Paris Saint-Germain2017 ~ 2022   
...                                               ...   
18974  Chongqing Dangdai Lifan FC SWM Team2020 ~ 2020   
18975                      Oldham Athletic2020 ~ 2021   
18976                           Wuhan Zall2018 ~ 2022   
18977                      Oldham Athletic2020 ~ 2021   
18978                         Club Bolívar2020 ~ 2024   

                                 Team_Name Contract_Year Start_Year End_Year  \
0                             FC Barcelona   2004 ~ 2021       2004     2021   
1                                 Juventus   2018 ~ 2022       2018     2022   
2                          Atlético Madrid   2014 ~ 2023       2014     2023   
3                          Manchester City   2015 ~ 2023       2015     2023   
4                      Paris Saint-Germain   2017 ~ 2022       2017     2022   
...                                    ...           ...        ...      ...   
18974  Chongqing Dangdai Lifan FC SWM Team   2020 ~ 2020       2020     2020   
18975                      Oldham Athletic   2020 ~ 2021       2020     2021   
18976                           Wuhan Zall   2018 ~ 2022       2018     2022   
18977                      Oldham Athletic   2020 ~ 2021       2020     2021   
18978                         Club Bolívar   2020 ~ 2024       2020     2024   

      Height  Height_cm  Weight  Weight_kg   foot  BOV   BP        Joined  \
0       5'7"     170.18  159lbs      72.12   Left   93   RW   Jul 1, 2004   
1       6'2"     187.96  183lbs      83.01  Right   92   ST  Jul 10, 2018   
2       6'2"     187.96  192lbs      87.09  Right   91   GK  Jul 16, 2014   
3      5'11"     180.34  154lbs      69.85  Right   91  CAM  Aug 30, 2015   
4       5'9"     175.26  150lbs      68.04  Right   91   LW   Aug 3, 2017   
...      ...        ...     ...        ...    ...  ...  ...           ...   
18974  5'10"     177.80  154lbs      69.85  Right   49   CB   Aug 1, 2020   
18975   5'7"     170.18  128lbs      58.06  Right   51  CAM   Aug 1, 2020   
18976  5'10"     177.80  146lbs      66.22  Right   49   CB  Jul 13, 2018   
18977   5'9"     175.26  143lbs      64.86  Right   51  CAM   Aug 1, 2020   
18978   5'9"     175.26  150lbs      68.04  Right   48  CDM   Jan 1, 2020   

        Value   Wage  
0      €67.5M  €560K  
1        €46M  €220K  
2        €75M  €125K  
3        €87M  €370K  
4        €90M  €270K  
...       ...    ...  
18974    €35K    €1K  
18975    €60K   €500  
18976    €40K    €1K  
18977    €60K   €500  
18978    €60K   €500  

[17727 rows x 21 columns]

In [23]:
# To convert the year columns into 'int' data type

df['Start_Year'] = df['Start_Year'].astype('int')
df['End_Year'] = df['End_Year'].astype('int')

# To verify that the year columns have been succesfully converted into 'int' data type
print(df['Start_Year'].mean())
print(df['End_Year'].max())

2018.3275793986575
2028


Nice, now we have succesfully converted the Start_Year and End_Year columns into integers, and can perform aggregrate functions on them.

### 9. Convert the value and wage (M- million, K- thousand) columns into int

In [24]:
# To convert the value column

df['Value_int'] = df['Value'].astype('str')
df['Value_int'] = df['Value'].str.replace('€','')


def convert_amount(value):
    if 'M' in value:
        return float(value.replace('M', '')) * 1e6
    elif 'K' in value:
        return float(value.replace('K', '')) * 1e3
    else:
        return float(value)

df['Value_int'] = df['Value_int'].apply(convert_amount)

df

Nationality Positions               Name  Age  ↓OVA  POT  \
0       Argentina  RW ST CF           L. Messi   33    93   93   
1        Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2        Slovenia        GK           J. Oblak   27    91   93   
3         Belgium    CAM CM       K. De Bruyne   29    91   91   
4          Brazil    LW CAM          Neymar Jr   28    91   91   
...           ...       ...                ...  ...   ...  ...   
18974    China PR        CB     Zhang Mengxuan   21    47   52   
18975     England        ST        V. Da Silva   17    47   67   
18976    China PR        CB             Xia Ao   21    47   55   
18977     England        CM           B. Hough   17    47   67   
18978     Bolivia       CDM          M. Flores   19    47   63   

                                      Team & Contract  \
0                             FC Barcelona2004 ~ 2021   
1                                 Juventus2018 ~ 2022   
2                          Atlético Madrid2014 ~ 2023   
3                          Manchester City2015 ~ 2023   
4                      Paris Saint-Germain2017 ~ 2022   
...                                               ...   
18974  Chongqing Dangdai Lifan FC SWM Team2020 ~ 2020   
18975                      Oldham Athletic2020 ~ 2021   
18976                           Wuhan Zall2018 ~ 2022   
18977                      Oldham Athletic2020 ~ 2021   
18978                         Club Bolívar2020 ~ 2024   

                                 Team_Name Contract_Year  Start_Year  \
0                             FC Barcelona   2004 ~ 2021        2004   
1                                 Juventus   2018 ~ 2022        2018   
2                          Atlético Madrid   2014 ~ 2023        2014   
3                          Manchester City   2015 ~ 2023        2015   
4                      Paris Saint-Germain   2017 ~ 2022        2017   
...                                    ...           ...         ...   
18974  Chongqing Dangdai Lifan FC SWM Team   2020 ~ 2020        2020   
18975                      Oldham Athletic   2020 ~ 2021        2020   
18976                           Wuhan Zall   2018 ~ 2022        2018   
18977                      Oldham Athletic   2020 ~ 2021        2020   
18978                         Club Bolívar   2020 ~ 2024        2020   

       End_Year Height  Height_cm  Weight  Weight_kg   foot  BOV   BP  \
0          2021   5'7"     170.18  159lbs      72.12   Left   93   RW   
1          2022   6'2"     187.96  183lbs      83.01  Right   92   ST   
2          2023   6'2"     187.96  192lbs      87.09  Right   91   GK   
3          2023  5'11"     180.34  154lbs      69.85  Right   91  CAM   
4          2022   5'9"     175.26  150lbs      68.04  Right   91   LW   
...         ...    ...        ...     ...        ...    ...  ...  ...   
18974      2020  5'10"     177.80  154lbs      69.85  Right   49   CB   
18975      2021   5'7"     170.18  128lbs      58.06  Right   51  CAM   
18976      2022  5'10"     177.80  146lbs      66.22  Right   49   CB   
18977      2021   5'9"     175.26  143lbs      64.86  Right   51  CAM   
18978      2024   5'9"     175.26  150lbs      68.04  Right   48  CDM   

             Joined   Value   Wage   Value_int  
0       Jul 1, 2004  €67.5M  €560K  67500000.0  
1      Jul 10, 2018    €46M  €220K  46000000.0  
2      Jul 16, 2014    €75M  €125K  75000000.0  
3      Aug 30, 2015    €87M  €370K  87000000.0  
4       Aug 3, 2017    €90M  €270K  90000000.0  
...             ...     ...    ...         ...  
18974   Aug 1, 2020    €35K    €1K     35000.0  
18975   Aug 1, 2020    €60K   €500     60000.0  
18976  Jul 13, 2018    €40K    €1K     40000.0  
18977   Aug 1, 2020    €60K   €500     60000.0  
18978   Jan 1, 2020    €60K   €500     60000.0  

[17727 rows x 22 columns]

In [25]:
# To convert the wage column

df['Wage_int'] = df['Wage'].astype('str')
df['Wage_int'] = df['Wage'].str.replace('€','')


def convert_amount(value):
    if 'M' in value:
        return float(value.replace('M', '')) * 1e6
    elif 'K' in value:
        return float(value.replace('K', '')) * 1e3
    else:
        return float(value)

df['Wage_int'] = df['Wage_int'].apply(convert_amount)

df

Nationality Positions               Name  Age  ↓OVA  POT  \
0       Argentina  RW ST CF           L. Messi   33    93   93   
1        Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2        Slovenia        GK           J. Oblak   27    91   93   
3         Belgium    CAM CM       K. De Bruyne   29    91   91   
4          Brazil    LW CAM          Neymar Jr   28    91   91   
...           ...       ...                ...  ...   ...  ...   
18974    China PR        CB     Zhang Mengxuan   21    47   52   
18975     England        ST        V. Da Silva   17    47   67   
18976    China PR        CB             Xia Ao   21    47   55   
18977     England        CM           B. Hough   17    47   67   
18978     Bolivia       CDM          M. Flores   19    47   63   

                                      Team & Contract  \
0                             FC Barcelona2004 ~ 2021   
1                                 Juventus2018 ~ 2022   
2                          Atlético Madrid2014 ~ 2023   
3                          Manchester City2015 ~ 2023   
4                      Paris Saint-Germain2017 ~ 2022   
...                                               ...   
18974  Chongqing Dangdai Lifan FC SWM Team2020 ~ 2020   
18975                      Oldham Athletic2020 ~ 2021   
18976                           Wuhan Zall2018 ~ 2022   
18977                      Oldham Athletic2020 ~ 2021   
18978                         Club Bolívar2020 ~ 2024   

                                 Team_Name Contract_Year  Start_Year  \
0                             FC Barcelona   2004 ~ 2021        2004   
1                                 Juventus   2018 ~ 2022        2018   
2                          Atlético Madrid   2014 ~ 2023        2014   
3                          Manchester City   2015 ~ 2023        2015   
4                      Paris Saint-Germain   2017 ~ 2022        2017   
...                                    ...           ...         ...   
18974  Chongqing Dangdai Lifan FC SWM Team   2020 ~ 2020        2020   
18975                      Oldham Athletic   2020 ~ 2021        2020   
18976                           Wuhan Zall   2018 ~ 2022        2018   
18977                      Oldham Athletic   2020 ~ 2021        2020   
18978                         Club Bolívar   2020 ~ 2024        2020   

       End_Year Height  Height_cm  Weight  Weight_kg   foot  BOV   BP  \
0          2021   5'7"     170.18  159lbs      72.12   Left   93   RW   
1          2022   6'2"     187.96  183lbs      83.01  Right   92   ST   
2          2023   6'2"     187.96  192lbs      87.09  Right   91   GK   
3          2023  5'11"     180.34  154lbs      69.85  Right   91  CAM   
4          2022   5'9"     175.26  150lbs      68.04  Right   91   LW   
...         ...    ...        ...     ...        ...    ...  ...  ...   
18974      2020  5'10"     177.80  154lbs      69.85  Right   49   CB   
18975      2021   5'7"     170.18  128lbs      58.06  Right   51  CAM   
18976      2022  5'10"     177.80  146lbs      66.22  Right   49   CB   
18977      2021   5'9"     175.26  143lbs      64.86  Right   51  CAM   
18978      2024   5'9"     175.26  150lbs      68.04  Right   48  CDM   

             Joined   Value   Wage   Value_int  Wage_int  
0       Jul 1, 2004  €67.5M  €560K  67500000.0  560000.0  
1      Jul 10, 2018    €46M  €220K  46000000.0  220000.0  
2      Jul 16, 2014    €75M  €125K  75000000.0  125000.0  
3      Aug 30, 2015    €87M  €370K  87000000.0  370000.0  
4       Aug 3, 2017    €90M  €270K  90000000.0  270000.0  
...             ...     ...    ...         ...       ...  
18974   Aug 1, 2020    €35K    €1K     35000.0    1000.0  
18975   Aug 1, 2020    €60K   €500     60000.0     500.0  
18976  Jul 13, 2018    €40K    €1K     40000.0    1000.0  
18977   Aug 1, 2020    €60K   €500     60000.0     500.0  
18978   Jan 1, 2020    €60K   €500     60000.0     500.0  

[17727 rows x 23 columns]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17727 entries, 0 to 18978
Data columns (total 23 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Nationality      17727 non-null  object 
 1   Positions        17727 non-null  object 
 2   Name             17727 non-null  object 
 3   Age              17727 non-null  int64  
 4   ↓OVA             17727 non-null  int64  
 5   POT              17727 non-null  int64  
 6   Team & Contract  17727 non-null  object 
 7   Team_Name        17727 non-null  object 
 8   Contract_Year    17727 non-null  object 
 9   Start_Year       17727 non-null  int64  
 10  End_Year         17727 non-null  int64  
 11  Height           17727 non-null  object 
 12  Height_cm        17727 non-null  float64
 13  Weight           17727 non-null  object 
 14  Weight_kg        17727 non-null  float64
 15  foot             17727 non-null  object 
 16  BOV              17727 non-null  int64  
 17  BP          

### 10. Finalize and remove unnecessary columns

In [27]:
df.drop(df.columns[[6,8,11,13,19,20]], inplace=True, axis=1) 

In [28]:
df

Nationality Positions               Name  Age  ↓OVA  POT  \
0       Argentina  RW ST CF           L. Messi   33    93   93   
1        Portugal     ST LW  Cristiano Ronaldo   35    92   92   
2        Slovenia        GK           J. Oblak   27    91   93   
3         Belgium    CAM CM       K. De Bruyne   29    91   91   
4          Brazil    LW CAM          Neymar Jr   28    91   91   
...           ...       ...                ...  ...   ...  ...   
18974    China PR        CB     Zhang Mengxuan   21    47   52   
18975     England        ST        V. Da Silva   17    47   67   
18976    China PR        CB             Xia Ao   21    47   55   
18977     England        CM           B. Hough   17    47   67   
18978     Bolivia       CDM          M. Flores   19    47   63   

                                 Team_Name  Start_Year  End_Year  Height_cm  \
0                             FC Barcelona        2004      2021     170.18   
1                                 Juventus        2018      2022     187.96   
2                          Atlético Madrid        2014      2023     187.96   
3                          Manchester City        2015      2023     180.34   
4                      Paris Saint-Germain        2017      2022     175.26   
...                                    ...         ...       ...        ...   
18974  Chongqing Dangdai Lifan FC SWM Team        2020      2020     177.80   
18975                      Oldham Athletic        2020      2021     170.18   
18976                           Wuhan Zall        2018      2022     177.80   
18977                      Oldham Athletic        2020      2021     175.26   
18978                         Club Bolívar        2020      2024     175.26   

       Weight_kg   foot  BOV   BP        Joined   Value_int  Wage_int  
0          72.12   Left   93   RW   Jul 1, 2004  67500000.0  560000.0  
1          83.01  Right   92   ST  Jul 10, 2018  46000000.0  220000.0  
2          87.09  Right   91   GK  Jul 16, 2014  75000000.0  125000.0  
3          69.85  Right   91  CAM  Aug 30, 2015  87000000.0  370000.0  
4          68.04  Right   91   LW   Aug 3, 2017  90000000.0  270000.0  
...          ...    ...  ...  ...           ...         ...       ...  
18974      69.85  Right   49   CB   Aug 1, 2020     35000.0    1000.0  
18975      58.06  Right   51  CAM   Aug 1, 2020     60000.0     500.0  
18976      66.22  Right   49   CB  Jul 13, 2018     40000.0    1000.0  
18977      64.86  Right   51  CAM   Aug 1, 2020     60000.0     500.0  
18978      68.04  Right   48  CDM   Jan 1, 2020     60000.0     500.0  

[17727 rows x 17 columns]

Perfect, the dirty data set started from having **18979 rows x 77 columns**, to now, a cleaned data set having **17727 rows x 17 columns** after we removed the unwanted data. We have succesfully cleaned the data as required and can proceed to the subsequent step! 

### 11. Export the cleaned data into .csv or .json file

In [ ]:
df.to_csv('fifa_cleaned.csv', index=False)

In [ ]:
df.to_json('fifa_cleaned.json')

The final step is exporting the cleaned data. We can export it into a file type that we desire, either **_.csv_** file or **_.json_** file. We have now completed the entire project, bravo!